zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/message/DbMessageCorrelationState.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.state.message;
+
+import io.camunda.zeebe.db.ColumnFamily;
+import io.camunda.zeebe.db.TransactionContext;
+import io.camunda.zeebe.db.ZeebeDb;
+import io.camunda.zeebe.db.impl.DbLong;
+import io.camunda.zeebe.engine.state.mutable.MutableMessageCorrelationState;
+import io.camunda.zeebe.protocol.ZbColumnFamilies;
+
+public class DbMessageCorrelationState implements MutableMessageCorrelationState {
+
+  private final DbLong messageKey;
+  private final RequestData requestData;
+
+  /** Message key -> Request data */
+  private final ColumnFamily<DbLong, RequestData> messageCorrelationColumnFamily;
+
+  public DbMessageCorrelationState(
+      final ZeebeDb<ZbColumnFamilies> zeebeDb, final TransactionContext transactionContext) {
+    messageKey = new DbLong();
+    requestData = new RequestData();
+    messageCorrelationColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.MESSAGE_CORRELATION, transactionContext, messageKey, requestData);
+  }
+
+  @Override
+  public void removeMessageCorrelation(final long messageKey) {
+    this.messageKey.wrapLong(messageKey);
+    messageCorrelationColumnFamily.deleteExisting(this.messageKey);
+  }
+
+  @Override
+  public void putMessageCorrelation(
+      final long messageKey, final long requestId, final int requestStreamId) {
+    this.messageKey.wrapLong(messageKey);
+    requestData.setRequestIdProp(requestId).setRequestStreamIdProp(requestStreamId);
+    messageCorrelationColumnFamily.insert(this.messageKey, requestData);
+  }
+
+  @Override
+  public RequestData getRequestData(final long messageKey) {
+    this.messageKey.wrapLong(messageKey);
+    return messageCorrelationColumnFamily.get(this.messageKey);
```

Comment:
❓ Should we return a copy of the record instead of the record itself ? Like we do for other records

❓ What happens if the request data is not found in the state ? Is it possible ?

Reply:
> Should we return a copy of the record instead of the record itself ? Like we do for other records

This isn't required, please note that the `RequestData` is not a record but a `DbValue`. We don't make copies for other `DbValues` either, e.g.:

```java
  @Override
  public StoredMessage getMessage(final long messageKey) {
    this.messageKey.wrapLong(messageKey);
    return messageColumnFamily.get(this.messageKey);
  }
```

> What happens if the request data is not found in the state ? Is it possible ?

This method will return `null` and we will likely end up in an NPE. This is why before getting it I check if it exists using the `existsRequestDataForMessageKey` method.
This does remind me that I missed to write a test for getting without value so I'll add that 😄

Reply:
> This isn't required, please note that the `RequestData` is not a record but a `DbValue`. We don't make copies for other `DbValues` either, e.g.:

This is not true:

```
  @Override
  public CompensationSubscription get(
      final String tenantId, final long processInstanceKey, final long recordKey) {
    wrapCompensationKeys(processInstanceKey, recordKey, tenantId);
    return compensationSubscriptionColumnFamily
        .get(tenantAwareProcessInstanceKeyCompensableActivityId)
        .copy();
  }
```

We had a discussion about this as a team [internal](https://camunda.slack.com/archives/C05DH1F5TAR/p1707908005239069) . `TransactionalColumnFamily.get` returns an object which gets mutated with every future get invocation

Reply:
I see. It's not a problem at this time as we won't get request data twice, but I'll change it to be future proof 👍